[RFC] drivers: stepper: initial support for stepper motors

gmarull opened the pull request:
A proposal to introduce a new API to control stepper motors. The API allows to control motor enablement, execute movements (sync and async) as well as set the speed. It has been designed to allow multiple stepper motors per controller.

The proposed implementation is for STM32 microcontrollers (tested on H7 only, but can be easily adjusted for all families). It uses two timers in master/slave mode (one to generate the steps and the second to keep track of the steps). It has been tested using a DRV8825 stepper driver. Note that while driver supports more than a single motor, only one can be active at the same time.

Example DT entry:
```
stepperctl0: {
	compatible = "st,stm32-stepperctl";

	#address-cells = <1>;
	#size-cells = <0>;
	label = "STEPPERCTL_0";
	master-timer = <&timers1>;
	slave-timer = <&timers2>;
        itr = <DT_STM32_TIMERS_TIM1_TIM2_ITR>;

	stepper0: stepper@2 {
		reg = <2>;
		microsteps = <1>;
		enable-gpios = <&gpioe 15 GPIO_ACTIVE_LOW>;
		direction-gpios = <&gpiod 8 GPIO_ACTIVE_HIGH>;
	};

	stepper1: stepper@1 {
		reg = <1>;
		microsteps = <8>;
		enable-gpios = <&gpioe 13 GPIO_ACTIVE_LOW>;
		direction-gpios = <&gpioe 12 GPIO_ACTIVE_HIGH>;
	};
};
```

Note that if this RFC is accepted I'll add documentation, examples and tests.

gmarull committed:
drivers: stepper: initial support for stepper motors

A proposal to introduce a new API to control stepper motors. The API
allows to control motor enablement, execute movements (sync and async)
as well as control the speed. It has been designed to allow multiple
stepper motors per controller.

The proposed implementation is for STM32 microcontrollers (tested on H7
only, but can be easily adjusted for all families). It uses two timers
in master/slave mode (one to generate the steps and the second to keep
track of the steps). It has been tested using a DRV8825 driver (step and
direction). Note that while driver supports more than a single motor,
only one can be active at the same time.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

dcpleung commented:
Q: what are the differences between sync and async movement? (like under what circumstances they are being used)

gmarull commented:
@dcpleung sync: waits until movement is completed, async: you are notified (useful e.g. to change the speed while moving)

dcpleung reviewed (approved):
API additions look good.

ABOSTM commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,757 @@
+	 *   the frequency of the APB domain.
+	 */
+	if (LL_RCC_GetTIMPrescaler() == LL_RCC_TIM_PRESCALER_TWICE) {
+		if (apb_psc == 1U) {
+			tim_clk = bus_clk;
+		} else {
+			tim_clk = bus_clk * 2U;
+		}
+	} else {
+		if (apb_psc == 1U || apb_psc == 2U || apb_psc == 4U) {
+			tim_clk = SystemCoreClock;
+		} else {
+			tim_clk = bus_clk * 4U;
+		}
+	}
+#else
+	if (pclken->bus == STM32_CLOCK_BUS_APB1) {
+		apb_psc = CONFIG_CLOCK_STM32_APB1_PRESCALER;
+	}
+#if !defined(CONFIG_SOC_SERIES_STM32F0X) && !defined(CONFIG_SOC_SERIES_STM32G0X)
+	else {
+		apb_psc = CONFIG_CLOCK_STM32_APB2_PRESCALER;
+	}
+#endif
+
+	/*
+	 * If the APB prescaler equals 1, the timer clock frequencies
+	 * are set to the same frequency as that of the APB domain.
+	 * Otherwise, they are set to twice (×2) the frequency of the
+	 * APB domain.
+	 */
+	if (apb_psc == 1U) {
+		tim_clk = bus_clk;
+	} else {
+		tim_clk = bus_clk * 2U;
+	}
+#endif
+
+	return tim_clk;
+}
```
This function get_timer_clock() doesn't match all STM32 series cases.
For example some stm32f4 socs or stm32f7 socs have also x1 and x4 cases.
You can have a look at Arduino implementation which cover most cases:
https://github.com/stm32duino/Arduino_Core_STM32/blob/73f76d6028f7171f6273aa3a52cb0a4117f92ff8/cores/arduino/HardwareTimer.cpp#L1253

gmarull replied:
Thanks, I'll take a look. PWM driver also needs a fix then, as this code is the same one.

ABOSTM commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,757 @@
+	}
+
+	if (steps == 0) {
+		LOG_ERR("Number of steps can not be zero");
```
why not? 
0 step -> do nothing.
Interest would be to have number of steps dynamically computed by application , and no need for application to check against 0.
Maybe necessary to tackle Asynchronous case.

ABOSTM reviewed (changes requested):
Thanks @gmarull for this PullRequest.

gmarull commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,757 @@
+	 * Duty cycle, d, is given by the ratio between the compare register
+	 * (ccr) and the auto-reload value (arr), that is:
+	 *
+	 * d = (ccr + 1) / (arr + 1).
```
Good spot, I missed the last point :-)

![image](https://user-images.githubusercontent.com/25011557/87333110-f2c61980-c53c-11ea-9f39-69beb03b3636.png)

NOTE: These ST slides should be fixed https://www.st.com/content/ccc/resource/training/technical/product_training/c4/1b/56/83/3a/a1/47/64/STM32L4_WDG_TIMERS_GPTIM.pdf/files/STM32L4_WDG_TIMERS_GPTIM.pdf/jcr:content/translations/en.STM32L4_WDG_TIMERS_GPTIM.pdf (slide 32)

gmarull commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,757 @@
+	 * yield to a value that does not fit into the register. In such case an
+	 * iterative computation is performed by changing psc.
+	 *
+	 * In order to prioritize runtime speed, the current psc value is used.
```
I have updated calculations, can you check? Really good suggestion, thanks.

gmarull commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,757 @@
+ * @param slave Slave timer.
+ * @param itr Where ITR will be stored.
+ */
+static int get_slave_timer_itr(TIM_TypeDef *master, TIM_TypeDef *slave,
```
Sounds reasonable, and at the same time these settings become available for other drivers if needed.

gmarull replied:
Added an initial version for H7. Can you check if this is what you guys expected? cc: @erwango

tbursztyka commented on drivers/stepper/Kconfig:
```diff
@@ -0,0 +1,25 @@
+# Copyright (c) 2020 Teslabs Engineering S.L.
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig STEPPER
```
Could you add [EXPERIMENTAL] ?

It's a really good API addition, but once it will be in master (hopefully for 2.4), users might have some improvements, changes etc... Could be a feature for position reference (homing) or else. 

Having it specifically set "experimental" will avoid to go through a tedious process of changing a stable API

gmarull replied:
good point, marked as experimental. thanks!

pabigot replied:
The API lifecycle doesn't require this, though perhaps it's something that could be added @carlescufi 

APIs start out as experimental and don't become stable until [certain conditions](https://docs.zephyrproject.org/latest/development_process/api_lifecycle.html#stable) are met including being available for two development releases.  A first step that seems to have been missed is adding this to the API project so it can be triaged and tracked.  I've done that.

tbursztyka replied:
Indeed, I am following the legacy rule. Which imo, could be kept as well. It's more obvious to users that the API is not stable if it's marked experimental in Kconfig.

carlescufi replied:
Yes, I will make a note to add this to the guide, thanks @pabigot.

pabigot replied:
Also to discuss this (Kconfig marking of EXPERIMENTAL) in API meeting.  It looks like a lot of Zephyr subsystems are doing this for features within  a subsystem, but possibly not for the entire API.  (Except for a couple like LoRa and Modem which aren't even listed in the API overview.)

gmarull commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,757 @@
+		}
+	} else {
+		if (apb_psc == 1U || apb_psc == 2U || apb_psc == 4U) {
+			tim_clk = SystemCoreClock;
```
I have fixed H7 code (this branch is only for H7).

gmarull committed:
dts: bindings: add stm32 h7 bindings for timers ITR

Timer internal trigger (ITR) settings have been added to DT bindings.
Only H7 series for now.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

ABOSTM commented on include/dt-bindings/timers/stm32_timers_h7.h:
```diff
@@ -0,0 +1,55 @@
+/* Helper macro to generate ITR (for master/slave in timers bits 4 and 5 of
+ * SMCR register).
+ */
+#define DT_STM32_TIMERS_ITR(n)		((n & 0x3U) << 4U)
```
This formula is valid only when ITR number is lower or equal to 3. 
But other series like STM32G4 can have ITR number up to 10.
And unfortunately, all bits of ITR are not adjacent one from each other (Bits 21, 20, 6, 5, 4 TS[4:0]: Trigger selection (see bits 21:20 for TS[4:3])).
I will discuss with @erwango on how to handle this issue with best genericity.

gmarull replied:
This header only applies to H7, where ITR goes from 0..3, so bits are adjacent. If some other series use non-adjacent bits the formula needs to take that into account.

gmarull replied:
@ABOSTM After looking at G4 series, I think that the idea of having a macro that converts a number to TS is not a good idea, as trigger selection number does not match the actual TS number. In case of G4, datasheet is confusing for (01000 ... 01110).

![image](https://user-images.githubusercontent.com/25011557/87654488-414f0000-c757-11ea-8b0c-e4c634623bf5.png)

gmarull replied:
@ABOSTM can you check latest changes wrt trigger selection in DT?

gmarull commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,653 @@
+	 * when using the full scale of arr (up to 65535). In such case,
+	 * the value of psc that will maximize the resolution is then:
+	 *
+	 * psc = ((f_tim / speed) / (arr + 1)) - 1
+	 *     = ((f_tim / speed) / 65536) - 1
+	 *     = ((f_tim / speed) >> 16) - 1.
```
Note that these calculations compute the register value (hence the -1). I can subtract the value when calling the LL prescaler set function if calculations become more clear.

ABOSTM replied:
Yes I notice that. But As I said, prescaler value is ((f_tim/speed) +1), then psc (register) is psc= Prescaler_value - 1
so   psc = (f_tim/speed) +1 -1 = f_tim/speed.
So I confirm -1 should be removed for the register. 
Imagine (f_tim/speed) is smaller than 65536, then due to entire division round down psc = -1  this is not good for register value !!!

gmarull replied:
Note that this calculation is not performed if arr fits into the 16-bit register, so psc = -1 is not possible (see https://github.com/zephyrproject-rtos/zephyr/pull/26221/files/b5615ec8eff0d67ed296a42e2e319ad185dbe992#diff-0321881e45a0440f0ae270f34324a9c6R377). In the docs `arr` and `psc` refer to actual register values, maybe I should clarify that.

gmarull commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,653 @@
+	arr = (data->f_tim / speed) - 1U;
+
+	if (!IS_TIM_32B_COUNTER_INSTANCE(cfg->master_timer) && arr > UINT16_MAX) {
+		psc = ((data->f_tim / speed) >> 16U) - 1U;
+		arr = UINT16_MAX;
```
I understand this way: we first have `f_tim / speed = K = (arr + 1) * (psc + 1)`. `K` is constant. Either arr or psc can be fixed. If I fix arr to its maximum value, `0xFFFF`, then psc is determined by `psc + 1 = K / (0xFFFF + 1) = K >> 16`. The advantage I see on fixing arr to its maximum value is to always have the maximum PWM resolution (not critical here, though). Am I missing something?

ABOSTM replied:
Your proposal give better dutycycle resolution (due to lower prescaler) but at a cost of inaccuracy on period (potentially very important) .
Lets take the worst example with your method.
 K = 0x1 FFFF
You fixed arr= 0xFFFF.
then psc = K / (0xFFFF +1) -1 = 0
But reverting the computation: with your method psc = 0 and arr = 0xFFFF 
then  (arr + 1) * (psc + 1) = 0x1000
Delta compare to expected K value DeltaK = 0x1 FFFF - 0x1000 = 0xFFFF
This is almost 50% erreur !!!
With your method, maximum inaccuracy on period is equal to 0xFFFF, because (0XFFFF+ 1) * (psc + 1) will take discrete values 0x10000; 0x20000; 0x30000 ... depending psc.
And there is still issue when K smaller than 0x10000 leading to psc =-1

With my method, I get psc =1 and arr = 0xFFFE
then  (arr + 1) * (psc + 1) = 0xFFFF * 2 = 0x1FFFE
Erreur on period is 1 .
With my method, max inaccuracy on period is equal to prescaler (if prescaler value is N, meaning N tick of tim input, then I may miss a maximum of N tick.)

gmarull replied:
Thanks @ABOSTM for the details. In this case period accuracy is the most important factor, so I see the point. I will fix the calculations, thanks again!

ABOSTM commented:
Edit: new proposed formula:
```
/* ITR selection may be composed of non adjacent bits :
 * TIM_SMCR_TS Bits 21, 20, 6, 5, 4 TS[4:0]: Trigger selection
 * (bit 20 and 21 are not available on all STM32 series)
 */
#define DT_STM32_TIMERS_ITR(n) ( ((n & 0x18U) << 20U) + (n & 0x7U) << 4U))
```

ABOSTM commented:
> @ABOSTM can you check latest changes wrt trigger selection in DT?

Yes, LGTM for DT trigger selection.

pabigot commented on include/drivers/stepper.h:
```diff
@@ -0,0 +1,140 @@
+
+static inline int z_impl_stepper_stop(struct device *dev, uint32_t channel)
+{
+	const struct stepper_driver_api *api =
```
Put all the declarations outside the #ifdef, and replace the body of this function with something like:
```
    int ret = -ENOTSUP;

#ifdef CONFIG_STEPPER_ASYNC
    const struct stepper_driver_api *api =
￼		(const struct stepper_driver_api *)dev->driver_api;

    if (api->stop) {
        ret = api->stop(dev, channel);
    }
#endif
   return ret;

pabigot commented on include/drivers/stepper.h:
```diff
@@ -0,0 +1,140 @@
+
+#ifdef CONFIG_STEPPER_ASYNC
+/**
+ * @brief Move stepper N steps (asynchronous API).
```
Since this API function is optional, we should also clarify whether drivers are required to implement it if it's enabled.

Part of the API discussion of this PR should include whether this *should* be optional, or required.

pabigot commented on include/drivers/stepper.h:
```diff
@@ -0,0 +1,140 @@
+{
+	const struct stepper_driver_api *api =
+		(const struct stepper_driver_api *)dev->driver_api;
+	return api->set_enabled(dev, channel, enabled);
```
empty line after declaration.  Multiple locations.

pabigot commented on include/drivers/stepper.h:
```diff
@@ -0,0 +1,140 @@
+ *
+ * @param dev	  Stepper controller device.
+ * @param channel Stepper channel.
+ * @param speed   Speed (steps/s).
```
If this can fail (e.g. the speed is not supported) the API should return `int` and document potential error codes.

What is the semantics of a negative speed? Direction?

What is the behavior on a zero speed?

pabigot commented on include/drivers/stepper.h:
```diff
@@ -0,0 +1,140 @@
+ * @param dev	  Stepper controller device.
+ * @param channel Stepper channel.
+ * @param steps   Number of steps.
+ */
```
What is the interpretation of the return value?  Negative should indicate an error; is there anything meaningful that could be represented by a positive value, or should zero be required as the success value?

pabigot commented on drivers/stepper/stepper_context.h:
```diff
@@ -0,0 +1,102 @@
+struct stepper_context {
+	struct k_sem lock;
+	struct k_sem sync;
+	int status;
```
How is `status` intended to be used?  What values go into it?

pabigot commented on drivers/stepper/stepper_context.h:
```diff
@@ -0,0 +1,102 @@
+	int status;
+
+#ifdef CONFIG_STEPPER_ASYNC
+	struct k_poll_signal *signal;
```
Can we eliminate the bool and use a non-null `signal` as a sign that the call is asynchronous?

pabigot commented on drivers/stepper/stepper_context.h:
```diff
@@ -0,0 +1,102 @@
+	k_sem_take(&ctx->lock, K_FOREVER);
+
+#ifdef CONFIG_STEPPER_ASYNC
+	ctx->asynchronous = asynchronous;
```
I guess this could be OK, but it'd be messy if `asynchronous` was true while `signal == NULL`.

pabigot commented on drivers/stepper/stepper_context.h:
```diff
@@ -0,0 +1,102 @@
+extern "C" {
+#endif
+
+struct stepper_context {
```
This `foo_context` header replicates a pattern that we know is wrong.  It supports asynchronous by adding a blocking async method, which is confusing.  It also uses a semaphore for coordination while a mutex may be more appropriate (not withstanding #25678).

I don't know that there's an alternative right now, but it might be wise to not provide an async API until we have a newer interface that addresses these concerns.

pabigot commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,653 @@
+	/** Pin. */
+	gpio_pin_t pin;
+	/** Flags. */
+	gpio_flags_t flags;
```
`gpio_dt_flags_t`.  It's smaller and would pack with `gpio_pin_t`.

pabigot commented on drivers/stepper/stepper_stm32.c:
```diff
@@ -0,0 +1,653 @@
+	return dev->config_info;
+}
+
+static inline int32_t get_motor_index(const struct stepper_stm32_config *config,
```
use `ssize_t` as the return type.

pabigot reviewed (changes requested):
A couple changes along with comments/questions about the API.

pabigot commented:
Capturing https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC5072_datasheet_Rev1.23.pdf as datasheet for a more complex motor control device.

carlescufi commented:
API meeting 21st July:

In general the API looks reasonable, with some caveats:

- The "channel" parameter needs to be clarified and discussed. Does it make sense in the context of the tree hierarchy? Instead an option would be to have the actual motors as DT nodes and then those refer to the timers that they share. In that case we would have multiple `device` structures, one per motor. Right now the controller is the node with each motor being defined inside it.

- Do we want to provide an API to control a single motor, or a generic one that can encapsulate stepper drivers that support multiple motors connected at a time?

- `set_enabled()`: this holds the torque, as opposed to running free, when not moving.

galak commented:
Add a link to one of the stepper motor control chips:

https://www.trinamic.com/fileadmin/assets/Products/ICs_Documents/TMC5072_datasheet_Rev1.23.pdf

gmarull commented:
Closing as I have no bandwidth to work on necessary changes. May re-open in the future if there is interest.

KenthJohan commented:
Any news on this? I used Zephyr once to control [TMC2130](https://www.trinamic.com/products/integrated-circuits/details/tmc2130/) using PWM for speed and GPIO for direction, but it also has I2C and so I ended up using Arduino to configure it with I2C because driver exists in Arduino environment. It would be nice to have driver support for these chips. Some configuration types might be unique to certain chips so a stepper driver with entity attribute value system to set arbitrary configurations could be useful. I think that drivers would spark interest for 3D printers to use Zephyr.
